Services/Component/classes/Setup/class.ilComponentInstallPluginObjective.php

```diff
@@ -0,0 +1,150 @@
+<?php
+/* Copyright (c) 2020 Daniel Weise <daniel.weise@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+declare(strict_types=1);
+
+use ILIAS\Setup;
+use ILIAS\DI;
+
+class ilComponentInstallPluginObjective implements Setup\Objective
+{
+    /**
+     * @var string
+     */
+    protected $plugin_name;
+
+    public function __construct(string $plugin_name)
+    {
+        $this->plugin_name = $plugin_name;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getHash() : string
+    {
+        return hash("sha256", self::class . $this->plugin_name);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getLabel() : string
+    {
+        return "Install plugin $this->plugin_name.";
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function isNotable() : bool
+    {
+        return true;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getPreconditions(Setup\Environment $environment) : array
+    {
+        $setup_config = $environment->getConfigFor('common');
+        $db_config = $environment->getConfigFor('database');
+
+        return [
+            new \ilIniFilesPopulatedObjective($setup_config),
+            new \ilDatabasePopulatedObjective($db_config),
+            new \ilComponentPluginAdminInitObjective()
+        ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function achieve(Setup\Environment $environment) : Setup\Environment
+    {
+        $ORIG_DIC = $this->initEnvironment($environment);
+
+        $plugin = $GLOBALS["DIC"]["ilPluginAdmin"]->getRawPluginDataFor($this->plugin_name);
+
+        if (!is_null($plugin) && $plugin['must_install'] && $plugin['supports_cli_setup']) {
+            $pl = ilPlugin::getPluginObject(
+                $plugin['component_type'],
+                $plugin['component_name'],
+                $plugin['slot_id'],
+                $plugin['name']
+            );
+
+            $pl->install();
+        }
+
+        $GLOBALS["DIC"] = $ORIG_DIC;
+
+        return $environment;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isApplicable(Setup\Environment $environment) : bool
+    {
+        $ORIG_DIC = $this->initEnvironment($environment);
+
+        $plugin = $GLOBALS["DIC"]["ilPluginAdmin"]->getRawPluginDataFor($this->plugin_name);
+
+        if (is_null($plugin) || !$plugin['supports_cli_setup']) {
+            return false;
+        }
+
+        $GLOBALS["DIC"] = $ORIG_DIC;
+
+        return $plugin['must_install'];
+    }
+
+    protected function initEnvironment(Setup\Environment $environment) : ILIAS\DI\Container
```

Comment:
Could we outsource this, since it exists at least three times? For example, as a trait?

Reply:
I think we could, but I don't think we should =)

I mean: From my point of view this is something we want to get rid of by using dependency injection a lot more. This snippet shows very much why we would want to do so. I would not bother to "improve" here by abstracting this away into some trait or something else.